packages/web-pkg/src/components/AppTemplates/AppWrapper.vue

```diff
@@ -0,0 +1,378 @@
+      })
+    }
+
+    const saveFileTask = useTask(function* () {
+      const newContent = unref(currentContent)
+      try {
+        const putFileContentsResponse = yield putFileContents(currentFileContext, {
+          content: newContent,
+          previousEntityTag: unref(currentETag)
+        })
+        serverContent.value = newContent
+        currentETag.value = putFileContentsResponse.etag
+      } catch (e) {
+        switch (e.statusCode) {
+          case 401:
+          case 403:
+            errorPopup(
+              new HttpError($gettext("You're not authorized to save this file"), e.response)
+            )
+            break
+          case 409:
+          case 412:
+            errorPopup(
+              new HttpError(
+                $gettext(
+                  'This file was updated outside this window. Please refresh the page (all changes will be lost).'
+                ),
+                e.response
+              )
+            )
+            break
+          case 507:
+            const space = store.getters['runtime/spaces/spaces'].find(
+              (space) => space.id === unref(resource).storageId && isProjectSpaceResource(space)
+            )
+            if (space) {
+              errorPopup(
+                new HttpError(
+                  $gettext('There is not enough quota on "%{spaceName}" to save this file', {
+                    spaceName: space.name
+                  }),
+                  e.response
+                )
+              )
+              break
+            }
+            errorPopup(
+              new HttpError($gettext('There is not enough quota to save this file'), e.response)
+            )
+            break
+          default:
+            errorPopup(new HttpError('', e.response))
+        }
+      }
+    }).restartable()
+
+    const save = async () => {
+      await saveFileTask.perform()
+    }
+
+    let autosaveIntervalId = null
+    onMounted(() => {
+      if (!unref(isEditor)) {
+        return
+      }
+      // Enable ctrl/cmd + s
+      document.addEventListener('keydown', handleSKey, false)
+      // Ensure reload is not possible if there are changes
+      window.addEventListener('beforeunload', handleUnload)
+      const editorOptions = store.getters.configuration.options.editor
+      if (editorOptions.autosaveEnabled) {
+        autosaveIntervalId = setInterval(async () => {
+          if (isDirty.value) {
+            await save()
+            autosavePopup()
+          }
+        }, (editorOptions.autosaveInterval || 120) * 1000)
+      }
+    })
+    onBeforeUnmount(() => {
+      if (!unref(isEditor)) {
+        return
+      }
+
+      window.removeEventListener('beforeunload', handleUnload)
+      document.removeEventListener('keydown', handleSKey, false)
+      clearInterval(autosaveIntervalId)
+      autosaveIntervalId = null
+    })
+
+    const handleSKey = function (e) {
+      if ((e.ctrlKey || e.metaKey) && e.code === 'KeyS') {
```

Comment:
@jacob-nv do we now have functions/constants we should use to handle this?

Reply:
You should be able to replace the whole listener with _useKeyboardActions_ composable. Just initialize it in setup and then create key binding. No need to worry about mounting, unmounting etc.
```    
// if you leave it null it will listen for document
const keyActions = useKeyboardActions('element-id')
// you will have to add S to enum (keycode) 
// the Modifier.Ctrl includes meta key
keyActions.bindKeyAction({ modifier: ModifierKey.Ctrl, primary: Key.S }, (event) => {   
// whatever code you need to execute
})
```

Reply:
Marking as resolved